### PR TITLE
Improve estimate gas and create access list rpcs

### LINKED ISSUE #434: EVM: Don't use `prepare_call_env` in `eth_estimateGas` and `eth_createAccessList`
that function was designed to create `TxEnv` for `eth_call`. We do a bunch of hotfixes to make it work with `eth_estimateGas`, don't need such a thing.

### PR DESCRIPTION
**Pull request by yaziciahmet**
# Description

## Linked Issues
- Fixes #434 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

### CODE REVIEW

File: crates/evm/src/evm/call.rs [hunk 1]
```diff
@@ -245,7 +236,7 @@ pub(crate) fn prepare_call_env(
             .map_err(|_| RpcInvalidTransactionError::GasUintOverflow)?,
         caller: from.unwrap_or_default(),
         gas_priority_fee: max_priority_fee_per_gas,
-        transact_to: to,
+        transact_to: to.unwrap_or_default(),
```
Comment (ercecan) [inline crates/evm/src/evm/call.rs:239]: good catch

File: crates/evm/src/evm/primitive_types.rs [hunk 2]
```diff
@@ -12,6 +12,7 @@ pub(crate) struct BlockEnv {
     /// basefee is added in EIP1559 London upgrade
     pub(crate) basefee: u64,
     pub(crate) gas_limit: u64,
+    pub(crate) l1_fee_rate: u128,
```
Comment (eyusufatik) [inline crates/evm/src/evm/primitive_types.rs:15]: if you do this, method id is going to change and we have to do this carefully with a fork.

pls revert this and releated changes :)

Comment (eyusufatik) [reply crates/evm/src/evm/primitive_types.rs:15]: it might be a smart change tho. make an issue and label with requires fork etc.

Comment (yaziciahmet) [reply crates/evm/src/evm/primitive_types.rs]: Reverted and created the issue #1266

**Review by kpp — APPROVED** [review]

**Review by ercecan — APPROVED** [review]

File: crates/evm/src/evm/handler.rs [hunk 3]
```diff
@@ -619,6 +619,7 @@ fn decrease_caller_balance<EXT, DB: Database>(
     change_balance(context, amount, false, address)
 }
 
+#[allow(unused)]
```
Comment (eyusufatik) [inline crates/evm/src/evm/handler.rs:622]: this should be used?

File: crates/evm/src/query.rs [hunk 4]
```diff
@@ -500,82 +499,51 @@ impl<C: sov_modules_api::Context> Evm<C> {
...
-            request.clone(),
-            Some(
-                evm_db
-                    .basic(request.from.unwrap_or_default())
-                    .unwrap()
-                    .unwrap_or_default()
-                    .balance,
-            ),
+            &mut cfg_env,
+            &mut request.clone(),
```
Comment (eyusufatik) [inline crates/evm/src/query.rs:539]: there might be a logical mistake here:

you are passing request as mutable reference, and in prepare_call_env, it is indeed mutated. 

however, you're passing the clone of request, which you keep using the original


my suggestion would be:


```rust
let cap_balance = evm_db.basic(request.from.unwrap_or_default()).....;

let tx_env = prepare_call_env(....., &mut request, cap_balance)?;
```

File: crates/evm/src/query.rs [hunk 5]
```diff
@@ -616,59 +584,51 @@ impl<C: sov_modules_api::Context> Evm<C> {
...
+            let block = self
+                .get_sealed_block_by_number(block_number, working_set)?
+                .ok_or(EthApiError::UnknownBlockNumber)?;
             let block_env = BlockEnv::from(&block);
 
-            (l1_fee_rate, block_env, block.header.number)
+            let cfg = self
+                .cfg
+                .get(working_set)
+                .expect("EVM chain config should be set");
```
Comment (eyusufatik) [inline crates/evm/src/query.rs:596]: reading config should happen after setting state to some block.

imagine you make a fork in the future, and change some value in config on the activation block.

with this setup, you are gonna run the latest cfg on a block from the past, which is possibly on an older version of the chain.

now your simulation will be wrong.

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: crates/evm/src/evm/handler.rs (hunk 3 above)
Comment (eyusufatik) [reply crates/evm/src/evm/handler.rs:622]: and yes, it is used.

in query.rs line 834

Comment (yaziciahmet) [reply crates/evm/src/evm/handler.rs:622]: it is not used in zk code, gives warning in build. was in between this and putting it behind native feature, chose unused.

File: crates/evm/src/query.rs (hunk 4 above)
Comment (yaziciahmet) [reply crates/evm/src/query.rs]: It is indeed funky, fixed it by making prepare_call_env consume transaction request

File: crates/evm/src/query.rs (hunk 5 above)
Comment (yaziciahmet) [reply crates/evm/src/query.rs:597]: fixed

File: crates/evm/src/evm/handler.rs (hunk 3 above)
Comment (eyusufatik) [reply crates/evm/src/evm/handler.rs:622]: native makes more sense actually. do it pls

Comment (yaziciahmet) [reply crates/evm/src/evm/handler.rs]: done

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**